### PR TITLE
Proposed change to make MMDMaterial more extensible

### PR DESCRIPTION
**Pull request by dshadowwolf**
This proposal makes it possible for other mods that base off MMDMaterial and MMDLib to more easily have custom blocks and items and custom material statistics by shifting the system to use a getter/setter style and have the storage of the data opaque to the user.

Review and comment - this major of a change needs reviewed and questions will arise - as will people spotting changes that need to be made.

### CODE REVIEW

File: src/main/java/com/mcmoddev/lib/material/MMDMaterial.java [hunk 1]
```diff
@@ -451,7 +388,127 @@ public MMDMaterial setBlastResistance(float resistance) {
...
+	 * @return the Block registered with the material, null if one of that name was not registered
+	 */
+	public Block getBlock( String name ) {
+		if( this.blocks.containsKey(name) ) {
+			return this.blocks.get(name);
+		}
+		return null;
+	}
+	
+	public boolean hasOre() {
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/material/MMDMaterial.java:457]: This is done wrong, as hasOre() is special:

hasOre should be set for non Oreless materials to *prevent* the creation of the ore block.

**Comment by Dizzy611** [comment]
Eyeballed it, am not familiar with surrounding codebase but the code itself looks solid and absent of syntax errors and major thinkos/typos.

File: src/main/java/com/mcmoddev/lib/material/MMDMaterial.java (hunk 1 above)
Comment (dshadowwolf) [reply src/main/java/com/mcmoddev/lib/material/MMDMaterial.java:457]: Fixed

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
+package com.mcmoddev.lib.data;
+
+public final class Names {
```
Comment (adrwa) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: This should be an enum imo.

File: src/main/java/com/mcmoddev/lib/data/MaterialStats.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package com.mcmoddev.lib.data;
+
+public final class MaterialStats {
```
Comment (adrwa) [inline src/main/java/com/mcmoddev/lib/data/MaterialStats.java]: Again, should be another enum

**Review by adrwa — CHANGES_REQUESTED** [review]

File: src/main/java/com/mcmoddev/basemetals/init/Blocks.java [hunk 4]
```diff
@@ -64,8 +65,8 @@ public static void init() {
 		if (Options.enableCoal) {
 			final MMDMaterial material = Materials.vanilla_coal;
 
-			material.block = net.minecraft.init.Blocks.COAL_BLOCK;
-			material.ore = net.minecraft.init.Blocks.COAL_ORE;
+			material.addNewBlock( Names.BLOCK, net.minecraft.init.Blocks.COAL_BLOCK );
```
Comment (adrwa) [inline src/main/java/com/mcmoddev/basemetals/init/Blocks.java:68]: Whats with this formatting?

File: src/main/java/com/mcmoddev/lib/material/MMDMaterial.java [hunk 5]
```diff
@@ -438,7 +378,10 @@ public int getTintColor() {
 	 * @return An instance of the material, for quality of life.
 	 */
 	public MMDMaterial setBlastResistance(float resistance) {
-		this.blastResistance = resistance;
+		if( this.stats.containsKey(MaterialStats.BLASTRESISTANCE) ) {
```
Comment (adrwa) [inline src/main/java/com/mcmoddev/lib/material/MMDMaterial.java]: Removing the key/value makes no difference, put will overwrite anyway

**Review by adrwa — CHANGES_REQUESTED** [review]

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 6]
```diff
@@ -1,77 +1,46 @@
...
-        public static final String ROD         = "rod";
-        public static final String SHEARS      = "shears";
-        public static final String SHIELD      = "shield";
-        public static final String SHOVEL      = "shovel";
-        public static final String SLAB        = "slab";
-        public static final String SMALLBLEND  = "smallblend";
-        public static final String SMALLPOWDER = "smallpowder";
-        public static final String SWORD       = "sword";
-        public static final String GEM         = "gem";
+	ANVIL,ARROW,AXE,BLEND,BOOTS,BOLT,BOW,CHESTPLATE,CRACKHAMMER,CROSSBOW,DOOR,FISHINGROD,GEAR,HELMET,HOE,HORSEARMOR,INGOT,LEGGINGS,NUGGET,PICKAXE,POWDER,ROD,SHEARS,SHIELD,SHOVEL,SLAB,SMALLBLEND,SMALLPOWDER,SWORD,GEM,
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: Formatting: Needs spaces

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 7]
```diff
@@ -1,77 +1,46 @@
...
-        public static final String TRAPDOOR      = "trapdoor";
-        public static final String TRIPWIREHOOK  = "tripwireHook";
-        public static final String WALL          = "wall";
-        public static final String FENCE         = "fence";
-        public static final String FENCEGATE     = "fenceGate";
-
-        // Fluid stuff
-        public static final String FLUID      = "fluid";
-        public static final String FLUIDBLOCK = "fluidBlock";
+	ANVILBLOCK,BARS,BLOCK,BOOKSHELF,BUTTON,DOORBLOCK,DOUBLESLAB,HALFSLAB,FLOWERPOT,LADDER,LEVER,PLATE,PRESSUREPLATE,STAIRS,TRAPDOOR,TRIPWIREHOOK,WALL,FENCE,FENCEGATE,
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: Formatting: needs spaces.

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 8]
```diff
@@ -1,77 +1,46 @@
...
 
         // Mekanism
-        public static final String CRYSTAL     = "crystal";
-        public static final String SHARD       = "shard";
-        public static final String CLUMP       = "clump";
-        public static final String POWDERDIRTY = "powderDirty";
+	CRYSTAL,SHARD,CLUMP,POWDERDIRTY,
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: Formatting: needs spaces.

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 9]
```diff
@@ -1,77 +1,46 @@
...
 
         // IC2
-        public static final String CASING          = "casing";
-        public static final String DENSEPLATE      = "densePlate";
-        public static final String CRUSHED         = "crushed";
-        public static final String CRUSHEDPURIFIED = "crushedPurified";
+	CASING,DENSEPLATE,CRUSHED,CRUSHEDPURIFIED,
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: Formatting: needs spaces.

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 10]
```diff
@@ -1,77 +1,46 @@
...
 
         // Ores
-        public static final String ORE       = "ore";
-        public static final String ENDORE    = "oreEnd";
-        public static final String NETHERORE = "oreNether";
+	ORE,ENDORE,NETHERORE;
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: Formatting: needs spaces.

File: src/main/java/com/mcmoddev/lib/data/Names.java [hunk 11]
```diff
@@ -1,77 +1,46 @@
...
+
+	
+    public static final TreeMap<Integer, Names> MAP = Maps.newTreeMap();
+
+    static {
+        for(Names name : values())
+            MAP.put(name.ordinal(),name);
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/Names.java]: Formatting: needs spaces.

File: src/main/java/com/mcmoddev/lib/data/MaterialStats.java [hunk 12]
```diff
@@ -1,29 +1,55 @@
...
 	 * The base attack damage for the material.
 	 */
-	public static final String BASEDAMAGE      = "baseDamage";
+	BASEDAMAGE;
+	
+    public static final TreeMap<Integer, MaterialStats> MAP = Maps.newTreeMap();
+
+    static {
+        for(MaterialStats stat : values())
+            MAP.put(stat.ordinal(),stat);
```
Comment (jriwanek) [inline src/main/java/com/mcmoddev/lib/data/MaterialStats.java]: Formatting: needs spaces.

**Review by jriwanek — CHANGES_REQUESTED** [review]

**Review by adrwa — APPROVED** [review]

File: src/main/java/com/mcmoddev/lib/data/MaterialStats.java (hunk 12 above)
Comment (jriwanek) [reply src/main/java/com/mcmoddev/lib/data/MaterialStats.java]: On  MAP.put(stat.ordinal(),stat);

not
        for(MaterialStats stat : values())

File: src/main/java/com/mcmoddev/lib/data/Names.java (hunk 11 above)
Comment (jriwanek) [reply src/main/java/com/mcmoddev/lib/data/Names.java]: On  MAP.put(stat.ordinal(),stat);

not
        for(MaterialStats stat : values())

**Review by jriwanek — APPROVED** [review]

**Pull request merged**